packages/astro/src/vite-plugin-astro/hmr.ts

```diff
@@ -22,6 +22,7 @@ export async function handleHotUpdate(
 	ctx: HmrContext,
 	{ config, logging, compile }: HandleHotUpdateOptions
 ) {
+	const importedModules = ctx.server.moduleGraph.getModuleById(ctx.file)?.importedModules ?? [];
```

Comment:
Weird hack! I discovered `ctx.modules` does _not_ contain any script tags on the page, but pulling directly from the module graph _will_ have these entries. Could be part of the problem.